File: lib/pull_preview/user_data.rb
module PullPreview
  class UserData
    attr_reader :app_path, :ssh_public_keys, :username

    def initialize(app_path:, ssh_public_keys:, username: "ec2-user")
      @app_path = app_path
      @ssh_public_keys = ssh_public_keys
      @username = username
    end

    def instructions
      result = []
      result << "#!/bin/bash"
      result << "set -xe ; set -o pipefail"
      if ssh_public_keys.any?
        result << %{echo '#{ssh_public_keys.join("\n")}' > /home/#{username}/.ssh/authorized_keys}
      end
      result << "mkdir -p #{app_path} && chown -R #{username}.#{username} #{app_path}"
      result << "echo 'cd #{app_path}' > /etc/profile.d/pullpreview.sh"
      result << "test -s /swapfile || ( fallocate -l 2G /swapfile && chmod 600 /swapfile && mkswap /swapfile && swapon /swapfile && echo '/swapfile none swap sw 0 0' | tee -a /etc/fstab )"

      # for Amazon Linux 2023, which by default creates a /tmp mount that is too small
      result << "systemctl disable --now tmp.mount"
      result << "systemctl mask tmp.mount"

      result << "sysctl vm.swappiness=10 && sysctl vm.vfs_cache_pressure=50"
      result << "echo 'vm.swappiness=10' | tee -a /etc/sysctl.conf"
      result << "echo 'vm.vfs_cache_pressure=50' | tee -a /etc/sysctl.conf"
      result << "yum install -y docker"
      result << %{curl -L "https://github.com/docker/compose/releases/download/v2.18.1/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose}
      result << "chmod +x /usr/local/bin/docker-compose"
      result << "usermod -aG docker #{username}"
      result << "systemctl restart docker"
      result << "echo 'docker system prune -f && docker image prune -a --filter=\"until=96h\" --force' > /etc/cron.daily/docker-prune && chmod a+x /etc/cron.daily/docker-prune"
      result << "mkdir -p /etc/pullpreview && touch /etc/pullpreview/ready && chown -R #{username}.#{username} /etc/pullpreview"
      result
    end

    def to_s
      instructions.join("\n")
    end
  end
end

File: internal/pullpreview/user_data.go
package pullpreview

import (
	"strings"
)

type UserData struct {
	AppPath       string
	SSHPublicKeys []string
	Username      string
}

func (u UserData) Script() string {
	homeDir := HomeDirForUser(u.Username)
	lines := []string{
		"#!/bin/bash",
		"set -xe ; set -o pipefail",
	}
	if len(u.SSHPublicKeys) > 0 {
		lines = append(lines, "echo '"+strings.Join(u.SSHPublicKeys, "\n")+"' > "+homeDir+"/.ssh/authorized_keys")
	}
	lines = append(lines,
		"mkdir -p "+u.AppPath+" && chown -R "+u.Username+":"+u.Username+" "+u.AppPath,
		"echo 'cd "+u.AppPath+"' > /etc/profile.d/pullpreview.sh",
		"test -s /swapfile || ( fallocate -l 2G /swapfile && chmod 600 /swapfile && mkswap /swapfile && swapon /swapfile && echo '/swapfile none swap sw 0 0' | tee -a /etc/fstab )",
		"systemctl disable --now tmp.mount",
		"systemctl mask tmp.mount",
		"sysctl vm.swappiness=10 && sysctl vm.vfs_cache_pressure=50",
		"echo 'vm.swappiness=10' | tee -a /etc/sysctl.conf",
		"echo 'vm.vfs_cache_pressure=50' | tee -a /etc/sysctl.conf",
		"yum install -y docker",
		"curl -L \"https://github.com/docker/compose/releases/download/v2.18.1/docker-compose-$(uname -s)-$(uname -m)\" -o /usr/local/bin/docker-compose",
		"chmod +x /usr/local/bin/docker-compose",
		"usermod -aG docker "+u.Username,
		"systemctl restart docker",
		"echo 'docker system prune -f && docker image prune -a --filter=\"until=96h\" --force' > /etc/cron.daily/docker-prune && chmod a+x /etc/cron.daily/docker-prune",
		"mkdir -p /etc/pullpreview && touch /etc/pullpreview/ready && chown -R "+u.Username+":"+u.Username+" /etc/pullpreview",
	)
	return strings.Join(lines, "\n")
}
